[2.7] Update to the dependencies guide (#2597)
astronomer/docs: learn/managing-dependencies.md

These statements are equivalent and result in the DAG shown in the following image: 

![List Dependencies](/img/guides/managing-dependencies_list_dependencies.png)

> When using bit-shift operators and the `.set_upstream` and `.set_downstream` method, it is not possible to set dependencies between two lists. For example, `[t0, t1] >> [t2, t3]` returns an error. To set dependencies between lists, use the dependency functions described in the next section.

Reviewer: ```suggestion
When you use bit-shift operators and the `.set_upstream` and `.set_downstream` method, you can't set dependencies between two lists. For example, `[t0, t1] >> [t2, t3]` returns an error. To set dependencies between lists, use the dependency functions described in the following section.
```

## Dependency functions

> Dependency functions are utilities that let you set dependencies between several tasks and lists of tasks. A common reason to use dependency functions over bit-shift operators is to create dependencies for tasks that were created in a loop and are stored in a list.

Reviewer: ```suggestion
Dependency functions are utilities that let you set dependencies between several tasks or lists of tasks. A common reason to use dependency functions over bit-shift operators is to create dependencies for tasks that were created in a loop and are stored in a list.
```

```python
from airflow.models.baseoperator import chain

list_of_tasks = []
for i in range(5):
    if i % 3 == 0:
        ta = EmptyOperator(task_id=f"ta_{i}")
        list_of_tasks.append(ta)
    else:
        ta = EmptyOperator(task_id=f"ta_{i}")
        tb = EmptyOperator(task_id=f"tb_{i}")
        tc = EmptyOperator(task_id=f"tc_{i}")
        list_of_tasks.append([ta, tb, tc])

chain(*list_of_tasks)
```

This code creates the following DAG structure:

![List Dependencies](/img/guides/managing-dependencies_chain_dependencies_1.png)

> ### chain()

Reviewer: ```suggestion
### Using `chain()`
```

To set parallel dependencies between tasks and lists of tasks of the same length, use the [`chain()` function](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/baseoperator/index.html#airflow.models.baseoperator.chain). For example:

```python
# from airflow.models.baseoperator import chain
chain(t0, t1, [t2, t3, t4], [t5, t6, t7], t8)
```

This code creates the following DAG structure:

![Chain Dependencies](/img/guides/managing-dependencies_chain_dependencies.png)

> When using the `chain` function, any lists or tuples that are set to depend directly on each other need to be of the same length.  

Reviewer: ```suggestion
When you use the `chain` function, any lists or tuples that are set to depend directly on each other need to be the same length.  
```

```python
chain([t0, t1], [t2, t3])  # this code will work
chain([t0, t1], [t2, t3, t4])  # this code will cause an error
chain([t0, t1], t2, [t3, t4, t5])  # this code will work
```

> ### chain_linear()

Reviewer: ```suggestion
### Using `chain_linear()`
```

To set interconnected dependencies between tasks and lists of tasks, use the `chain_linear()` function. This function is available in Airflow 2.7+, in older versions of Airflow you can set similar dependencies between two lists at a time using the [`cross_downstream()` function](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/baseoperator/index.html#airflow.models.baseoperator.cross_downstream). 

> Replacing `chain` in the previous example with `chain_linear` will create dependencies such that each element in the downstream list will depend on each element in the upstream list.

Reviewer: ```suggestion
Replacing `chain` in the previous example with `chain_linear` creates dependencies where each element in the downstream list will depend on each element in the upstream list.
```

```python
# from airflow.models.baseoperator import chain_linear
chain_linear(t0, t1, [t2, t3, t4], [t5, t6, t7], t8)
```

![Chain Linear Dependencies 2](/img/guides/managing-dependencies_chain_linear_dependencies_1.png)

> The `chain_linear()` function can take in lists of any length in any order. For example:

Reviewer: ```suggestion
The `chain_linear()` function can accept lists of any length in any order. For example, the following arguments are valid:
```

```python
chain_linear([t0, t1], [t2, t3, t4])
```

![Chain Linear Dependencies 1](/img/guides/managing-dependencies_chain_linear_dependencies_2.png)

## Dependencies in dynamic task mapping

> Dependencies for [dynamically mapped tasks](dynamic-tasks.md) can be set in the same way as regular tasks. Keep in mind that when using the default [trigger rule](#trigger-rules) `all_success`, all mapped task instances need to be successful for the downstream task to run. For the purpose of trigger rules, mapped task instances behave like a set of parallel upstream tasks.

Reviewer: ```suggestion
Dependencies for [dynamically mapped tasks](dynamic-tasks.md) can be set in the same way as regular tasks. Note that when using the default [trigger rule](#trigger-rules) `all_success`, all mapped task instances need to be successful for the downstream task to run. For the purpose of trigger rules, mapped task instances behave like a set of parallel upstream tasks.
```

# end will only run if all mapped task instances of the multiply task are successful
end = EmptyOperator(task_id="end")

start >> multiply_obj >> end

# all of the following ways of setting dependencies are valid
# multiply_obj.set_downstream(end)
# end.set_upstream(multiply_obj)
# chain(start, multiply_obj, end)

```

![Dependencies dynamic tasks](/img/guides/managing-dependencies_dynamic_tasks_decorator.png)

</TabItem>

<TabItem value="traditional">

```python
start=EmptyOperator(task_id="start")

def multiply_func(x,y):
    return x*y

multiply_obj = PythonOperator.partial(
    task_id="multiply",
    python_callable=multiply_func,
    op_args=[2]
).expand(op_kwargs=[{"y": 1}, {"y": 2}, {"y": 3}])

# end will only run if all mapped task instances of the multiply task are successful
end = EmptyOperator(task_id="end")

start >> multiply_obj >> end

# all of the following ways of setting dependencies are valid
# multiply_obj.set_downstream(end)
# end.set_upstream(multiply_obj)
# chain(start, multiply_obj, end)

```

![Dependencies dynamic tasks](/img/guides/managing-dependencies_dynamic_tasks_traditional.png)

</TabItem>
</Tabs>

## Task group dependencies

> [Task groups](task-groups.md) are an Airflow concept allowing for grouping of tasks in the UI and for [dynamic mapping use cases](task-groups.md#generate-task-groups-dynamically-at-runtime). You can get more information on task groups in the respective guide. This section will explain how to set dependencies between task groups.

Reviewer: ```suggestion
[Task groups](task-groups.md) are tasks that are logically grouped in the Airflow UI and for [dynamic mapping use cases](task-groups.md#generate-task-groups-dynamically-at-runtime). This section will explain how to set dependencies between task groups.
```

> Dependencies can be set both inside and outside of a task group. For example, in the following DAG code there is a start task, a task group with two dependent tasks, and an end task. All of these task need to happen sequentially. The dependencies between the two tasks in the task group are set within the task group's context (`t1 >> t2`). The dependencies between the task group and the start and end tasks are set within the DAG's context (`t0 >> tg1() >> t3`).

Reviewer: ```suggestion
Dependencies can be set both inside and outside of a task group. For example, in the following DAG code there is a start task, a task group with two dependent tasks, and an end task. All of these tasks need to happen sequentially. The dependencies between the two tasks in the task group are set within the task group's context (`t1 >> t2`). The dependencies between the task group and the start and end tasks are set within the DAG's context (`t0 >> tg1() >> t3`).
```

These statements are equivalent and result in the DAG shown in the following image: 

![List Dependencies](/img/guides/managing-dependencies_list_dependencies.png)

When using bit-shift operators and the `.set_upstream` and `.set_downstream` method, it is not possible to set dependencies between two lists. For example, `[t0, t1] >> [t2, t3]` returns an error. To set dependencies between lists, use the dependency functions described in the next section.

## Dependency functions

> Dependency functions are utilities that let you set dependencies between several tasks and lists of tasks. A common reason to use dependency functions over bit-shift operators is for example to create dependencies for tasks that were created in a loop and are stored in a list.

Reviewer 2: ```suggestion
Dependency functions are utilities that let you set dependencies between several tasks and lists of tasks. A common reason to use dependency functions over bit-shift operators is to create dependencies for tasks that were created in a loop and are stored in a list.
```

```python
# from airflow.models.baseoperator import chain
chain(t0, t1, [t2, t3, t4], [t5, t6, t7], t8)
```

This code creates the following DAG structure:

![Chain Dependencies](/img/guides/managing-dependencies_chain_dependencies.png)

When using the `chain` function, any lists or tuples that are set to depend directly on each other need to be of the same length.  

```python
chain([t0, t1], [t2, t3])  # this code will work
chain([t0, t1], [t2, t3, t4])  # this code will cause an error
chain([t0, t1], t2, [t3, t4, t5])  # this code will work
```

### chain_linear()

To set interconnected dependencies between tasks and lists of tasks, use the `chain_linear()` function. This function is available in Airflow 2.7+, in older versions of Airflow you can set similar dependencies between two lists at a time using the [`cross_downstream()` function](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/baseoperator/index.html#airflow.models.baseoperator.cross_downstream). 

> Replacing `chain` with `chain_linear` in the previous example as shown in this code snippet will create a DAG structure where the dependency between lists is set in a pattern so that each element in the downstream list will depend on each element in the upstream list.

Reviewer 2: ```suggestion
Replacing `chain` in the previous example with `chain_linear` will create dependencies such that each element in the downstream list will depend on each element in the upstream list.
```

```python
# from airflow.models.baseoperator import chain_linear
chain_linear(t0, t1, [t2, t3, t4], [t5, t6, t7], t8)
```

![Chain Linear Dependencies 2](/img/guides/managing-dependencies_chain_linear_dependencies_1.png)

The `chain_linear()` function can take in lists of any length in any order. For example:

```python
chain_linear([t0, t1], [t2, t3, t4])
```

![Chain Linear Dependencies 1](/img/guides/managing-dependencies_chain_linear_dependencies_2.png)

> ## Loops and dependencies

Reviewer 2: High level, do we think generating tasks in a loop is still relevant now that dynamic task mapping exists? I'm sure there are edge cases, but is it relevant enough to keep this section? I would lean towards removing it, but open to arguments otherwise.

Author: Hmm... I think using loops to create tasks is probably still used a lot for non-dynamic mapping (e.g. I have this task and 20 fixed tables it needs to run on) but yes we should encourage best practices and I can't really think of a case (now that dynamic task group mapping exists) where looping would have any advantage over using dynamic mapping 🤔 
I mean, unless people are on old versions, that would be my one weak argument for keeping it. 
In general I think if someone is using a loop to create tasks they probably also know how to handle objects in a loop and be able to figure out the info in this section. 

I will remove the section for now but save it somewhere just in case. 😄

If you generate tasks in a loop in your DAG, you can define the dependencies within the context of the code used to create the tasks.

In the following example, a set of parallel tasks is generated by looping through a list of endpoints. Each `generate_files` task is downstream of `start` and upstream of `send_email`.

```python
t0 = EmptyOperator(task_id="start")

send_email = EmailOperator(
    task_id="send_email",
    to=email_to,
    subject="Data to S3 DAG",
    html_content="<p>The data to S3 DAG completed successfully. \
    Files can now be found on S3. <p>",
)

for endpoint in endpoints:
    generate_files = PythonOperator(
        task_id="generate_file_{0}".format(endpoint),
        python_callable=upload_to_s3,
        op_kwargs={"endpoint": endpoint, "date": date},
    )

    t0 >> generate_files >> send_email

```

This image shows the resulting DAG:

![Loop Dependencies](/img/guides/managing-dependencies_loop_dep.png)

## Dependencies in dynamic task mapping

> Tasks that utilize [dynamic task mapping](dynamic-tasks.md) can be set to be up- and downstream of other tasks as with regular tasks. Keep in mind that when using the default [trigger rule](#trigger-rules) `all_success`, all mapped task instances need to be successful for the downstream task to run. For the purpose of trigger rules, mapped task instances behave like a set of parallel upstream tasks.

Reviewer 2: ```suggestion
Dependencies for [dynamically mapped tasks](dynamic-tasks.md) can be set in the same way as regular tasks. Keep in mind that when using the default [trigger rule](#trigger-rules) `all_success`, all mapped task instances need to be successful for the downstream task to run. For the purpose of trigger rules, mapped task instances behave like a set of parallel upstream tasks.
```

This image shows the resulting DAG:

![Task Group Dependencies](/img/guides/managing-dependencies_tg_dependencies_1.png)

> You can also set dependencies between task groups, between tasks outside of a task group and tasks inside a task group and even between tasks in different (nested) task groups.

Reviewer: ```suggestion
You can also set dependencies between task groups, between tasks inside and out of task groups, and even between tasks in different (nested) task groups.
```

![Task Group Dependencies](/img/guides/managing-dependencies_task_group_complex.png)

## TaskFlow API dependencies

The [TaskFlow API](airflow-decorators.md) `@task` decorator allows you to easily turn Python functions into Airflow tasks. 

If your DAG has several tasks that are defined with the `@task` decorator and use each other's output, you can leverage inferred dependencies via the TaskFlow API. For example, in the following DAG there are two dependent tasks, `get_a_cat_fact` and `print_the_cat_fact`. To set the dependencies, you pass the called function of the upstream task as a positional argument to the downstream task (`print_the_cat_fact(get_a_cat_fact())`):

<CodeBlock language="python">{dependencies_example_2_taskflow}</CodeBlock>

This image shows the resulting DAG:

![TaskFlow Dependencies](/img/guides/managing-dependencies_taskflow_1.png)

> Note that it is also possible to assign the called function to an object and then pass that object to the downstream task. This way of defining dependencies between `@task` decorated tasks is often easier to read and allows you to set the same task as an upstream dependency to multiple other tasks. 

Reviewer: ```suggestion
Note that you can also assign the called function to an object and then pass that object to the downstream task. This way of defining dependencies is often easier to read and allows you to set the same task as an upstream dependency to multiple other tasks. 
```

This image shows the resulting DAG:

![Task Group Dependencies](/img/guides/managing-dependencies_tg_dependencies_1.png)

You can also set dependencies between task groups, between tasks outside of a task group and tasks inside a task group and even between tasks in different (nested) task groups.

The image below shows types of dependencies that can be set between tasks and task groups. You can find the code that created this DAG in a GitHub repository both for the [TaskFlow API](https://github.com/astronomer/webinar-task-groups/blob/main/dags/example_complex_dependencies_1.py) and [traditional version](https://github.com/astronomer/webinar-task-groups/blob/main/dags/example_complex_dependencies_2.py).

![Task Group Dependencies](/img/guides/managing-dependencies_task_group_complex.png)

## TaskFlow API dependencies

> The [TaskFlow API](airflow-decorators.md) added decorators to Airflow, one of these decorators, `@task`, allows you to easily turn Python functions into Airflow tasks. 

Reviewer 2: ```suggestion
The [TaskFlow API](airflow-decorators.md) `@task` decorator allows you to easily turn Python functions into Airflow tasks. 
```
